File: backend/app/api/v1/compare.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.domain.compare import CompareResponse, CompareService
from app.domain.versions.repository import VersionRepository

router = APIRouter()


@router.get("/documents/{document_id}/compare", response_model=CompareResponse)
def compare_versions(
    document_id: str,
    from_version: str,
    to_version: str,
    db: Session = Depends(get_db),
) -> CompareResponse:
    compare_service = CompareService(VersionRepository(db))
    try:
        return compare_service.compare(document_id, from_version, to_version)
    except ValueError as error:
        raise HTTPException(status_code=404, detail=str(error)) from error

File: frontend/src/lib/api/compare.ts
import { apiRequest } from "@/lib/api/client";

export type CompareResponse = {
  document_id: string;
  from_version: string;
  to_version: string;
  changed_pages: number[];
};

export async function compareVersions(documentId: string, fromVersion: string, toVersion: string) {
  const params = new URLSearchParams({ from_version: fromVersion, to_version: toVersion });
  return apiRequest<CompareResponse>(`/documents/${documentId}/compare?${params.toString()}`);
}
